### ISSUE TITLE
Heartbeat for Publisher and ClusterRPCProxy

### ISSUE THREAD

**Issue by bobh66** [nameko/nameko#562]
I have several long-running services that invoke Nameko services by ClusterRPCProxy, or they send events into Nameko services using the event_dispatcher.

I noticed that the RabbitMQ timeout/hearbeat for the Nameko services is always 60, but the connections from the ClusterRPCProxy and event Publisher have the heartbeat set to 0.

Is there any reason not to enable the default heartbeat on the Connection objects created by the Publisher and PollingQueueConsumer objects?  It can be configurable like the ssl option that was recently added.

I'm happy to submit a PR if this is acceptable.

Thanks

**Comment by mattbennett** [comment]
Heartbeats aren't supported on publisher connections because there's no easy way (in kombu) to periodically send the heartbeat. Unlike the consumer, there is no thread or loop around which we're constantly iterating, thus no opportunity to send the heartbeat.

This is one of the many reasons why my long-standing ambition is to replace kombu with a different AMQP library. The AMQP extensions have been refactored significantly recently (especially in https://github.com/nameko/nameko/pull/542) which makes this reality a bit more likely, but it'd still be quite a lot of work.

**Comment by s-maj** [comment]
Instead of using "tricky" AMQP heartbeats there is an easier solution: [TCP keepalives](https://www.rabbitmq.com/heartbeats.html#tcp-keepalives). This feature was initially introduced in [py-amqp 2.0](http://amqp.readthedocs.io/en/latest/changelog.html#version-2-0-0) and tuned-up in [py-amqp 2.1.4](https://amqp.readthedocs.io/en/latest/changelog.html#version-2-1-4) (it's not supported by librabbitmq but so the AMQP heartbeats). Sadly, this is blocked by `kombu<4` dependency.

**Comment by mattbennett** [comment]
Nice! TCP keepalives would be a much easier solution.

Nameko pins `kombu<4` because of the jump to pyamqp 2.x, which is not backwards compatible with the 1.x series. The incompatibility is small though, and I think we're ready make the leap and drop support for the older versions now.

From memory, you can no longer declare on a `Connection` and must use the `Channel` instead, and there are more explicit exceptions (which I think make detecting undeliverable messages easier)

**Comment by s-maj** [comment]
Ok, this would be extremely useful for deployments with RabbitMQ clustered installations. In case of abrupt stop (hardware failure, network failure, etc) one of nodes in RabbitMQ cluster, producer is currently unable to detect this and reconnect to the healthy node. This also should unblock folks from https://github.com/nameko/nameko/issues/378.

I did a quick PoC according to your suggestions (https://github.com/nameko/nameko/pull/564) but some tests are failing.

**Comment by mattbennett** [comment]
TCP keepalives supported with the changes in #564

**Comment by bobh66** [comment]
@s-maj Thanks for all of the effort!  The original investigation was triggered by RabbitMQ holding open sockets that had been closed (maybe improperly) by the client, so when a publisher re-used the socket (because it was available on the client side) it hung for 15+ minutes since the server thought the socket was already open.  I haven't (yet) been able to capture a trace of the client socket closure that leaves the server socket open, but I'll test with 2.11 when it's available and see if the TCP keepalives prevent the hung connections.

Thanks again!

**Comment by s-maj** [comment]
Ok, cool! This should help as long you have direct communication (no NAT, not load balancer). You could also configure aggressive TCP keepalive on the RabbitMQ server(s) - default value timeout value is over 2h.

**Comment by bobh66** [comment]
TCP keepalive on the server doesn't seem to help, even with the default 2 hour timeout I have found connections that have been stuck open for weeks, according to the connected_at time reported by rabbitmqctl list_connections.

Since the client side is originating the TCP connection, I wasn't sure if the server-side TCP keepalive was invoked.  Client-side TCP keepalive might not help if the socket is being torn down improperly without completing the AMQP disconnect procedure that RabbitMQ is looking for.

I added AMQP heartbeats to the Publisher and ClusterRPCProxy connections and that seemed to help, but it also causes the connections to be torn down quickly since it is not responding to the heartbeat from the server, so it prevents reusing connections from the kombu pool.
